### 데이터 소스 : titanic_df = sns.load_dataset('titanic')
### Q. embark_town 열의 829행의 NaN 데이터의 앞, 뒤를 포함하는 3개행을 출력하세요.

In [15]:
import seaborn as sns

titanic_df = sns.load_dataset('titanic')

print(titanic_df.embark_town.iloc[828:831])

828    Queenstown
829           NaN
830     Cherbourg
Name: embark_town, dtype: object


### Q. titanic_df embark_town 열의 NaN값을 바로 앞에 있는 830행의 값으로 변경한 후 출력하세요.

In [57]:
titanic_df['embark_town'].fillna(method='bfill', inplace=True)

print(titanic_df.embark_town.iloc[828:831])

828    Queenstown
829     Cherbourg
830     Cherbourg
Name: embark_town, dtype: object


### Q. embark_town 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 대체하세요.

In [20]:
embark_max = titanic_df.embark_town.value_counts().idxmax()
print()

titanic_df['embark_town'].fillna(embark_max, inplace=True)
print(titanic_df['embark_town'].isnull().sum(axis=0))


0


### Q. titanic_df에서 'age' 열에 나이 데이터가 없는 모든 행을 삭제 후 df_age 이름으로 저장하고 행의 갯수를 구하세요.  데이터 소스 : titanic_df = sns.load_dataset('titanic')

In [5]:
import seaborn as sns
import pandas as pd

titanic_df = sns.load_dataset('titanic')

# titanic_df.head()
# print()
# titanic_df.age.notnull()

df_age = titanic_df['age'].dropna(axis=0)
print(df_age.value_counts().sum())

714


### Q. age 열의 NaN값을 다른 나이 데이터의 평균으로 변경한 후 'age' 열의 첫 10개 데이터만 정수로 출력하세요.

In [6]:
age_ava = titanic_df.age.sum() / titanic_df.age.value_counts().sum()
titanic_df['age'].fillna(age_ava, inplace=True)

print(titanic_df['age'].astype(int).head(10))
print()
print(titanic_df['age'].isnull().sum())

0    22
1    38
2    26
3    35
4    35
5    29
6    54
7     2
8    27
9    14
Name: age, dtype: int32

0


In [7]:
mean_age = titanic_df['age'].mean(axis=0)
titanic_df['age'].fillna(mean_age, inplace=True)
print(titanic_df['age'].astype('int').head(10))

0    22
1    38
2    26
3    35
4    35
5    29
6    54
7     2
8    27
9    14
Name: age, dtype: int32


### Q. 1 ~ 2사이의 정수로 (5,3) 2차원 배열을 랜덤하게 만든 후 DataFrame으로 불러와 df에 저장하고 df에서 중복 행을 제거한 후 df2에 저장 후 출력하세요. (columns = ['a', 'b', 'c'])

In [9]:
import numpy as np
import pandas as pd

columns = ['a', 'b', 'c']
df = pd.DataFrame(np.random.randint(1, 3, size=(5,3)), columns=columns)

display(df)

df2 = df.drop_duplicates()

display(df2)

a  b  c
0  1  2  1
1  1  2  2
2  2  2  2
3  1  1  1
4  2  2  1

a  b  c
0  1  2  1
1  1  2  2
2  2  2  2
3  1  1  1
4  2  2  1

### Q. df에서 b, c열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.

In [10]:
df3 = df[['b', 'c']].drop_duplicates()

display(df3)

b  c
0  2  1
1  2  2
3  1  1

In [11]:
df3 = df.drop_duplicates(subset=['a','b'])
display(df3)

a  b  c
0  1  2  1
2  2  2  2
3  1  1  1

In [56]:
# 데이터 프레임 생성
import pandas as pd
auto_df = pd.read_csv('auto-mpg.csv', header=None)
auto_df.columns = ['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name'] 
 
auto_df.head(3)

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite

### Q. horsepower 열의 누락 데이터 '?'을 삭제한 후 NaN 값의 갯수를 출력하세요.

In [61]:
import numpy as np
auto_df['horsepower'].replace('?', np.nan, inplace=True)
auto_df.dropna(subset=['horsepower'], axis=0, inplace=True)
auto_df['horsepower'].isnull().sum()
print(auto_df.horsepower.dtypes)
print(auto_df.horsepower.isnull().sum())

object
0


### Q. horsepower 열의 누락 데이터('?') 삭제하고 실수형으로 변환하세요.

In [64]:
auto_df['horsepower'].replace('?', np.nan, inplace=True)
auto_df.dropna(subset=['horsepower'], axis=0, inplace=True)
auto_df['horsepower'].isnull().sum()
auto_df['horsepower'] = auto_df['horsepower'].astype('float')

print(auto_df['horsepower'].dtypes)
print(auto_df.head(3))

float64
    mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   

   model year  origin                       name  
0          70       1  chevrolet chevelle malibu  
1          70       1          buick skylark 320  
2          70       1         plymouth satellite  


### Q. Titanic 가공한 내역을 정리하고 이를 쉽게 재사용할 후 있도록 종합 전처리 함수를 만드세요.(def transform_features(df):)

In [ ]:
from sklearn.preprocessing import LabelEncoder

# null 처리 함수
def fillna(df):
    df['Age'].fillna(df['Age'].mean(), inplace=True)
    df['cabin'].fillna('N', inplace=True)
    df['Embarked'].fillna('N', inplace=True)
    return df

# 불필요 속성 제거
def drop_features(df):
    df.drop(['Passengerid', 'Name', 'Ticket'], axis=1, inplace=True)
    return df

# 레이블 인코딩
def format_features(df):
    df['Cabin'] = df['Cabin'].str[:1]
    features = ['Cabin', 'Sex', 'Embarked']
    for feature in features:
        le = LabelEncoder()
        le = le.fit(df[feature])
        df[feature] = le.transform(df[feature])
    return df

In [ ]:
# 앞에서 설정한 Data Preprocessing 함수 호출
def transform_features(df):
    df = fillna(df)
    df = drop_features(df)
    df = format_features(df)
    return df

In [ ]:
import pandas as pd
titanic_df = pd.read_csv('train.csv')

transform_features(titanic_df)

In [14]:
import pandas as pd
titanic_df = pd.read_csv('train.csv')

In [15]:
titanic_df.drop(['PassengerId', 'Name', 'Ticket'], axis=1, inplace=True)
print(titanic_df.columns.values)
print(titanic_df.columns)

['Survived' 'Pclass' 'Sex' 'Age' 'SibSp' 'Parch' 'Fare' 'Cabin' 'Embarked']
Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin',
       'Embarked'],
      dtype='object')


In [16]:
titanic_df.Age
titanic_df.Age.isnull().sum()
titanic_df.Age.mean()
titanic_df.Age.fillna(titanic_df.Age.mean(), inplace=True)
titanic_df.Age.isnull().sum()

0

In [20]:
titanic_df.Cabin
titanic_df.Cabin.isnull().sum()
titanic_df.Cabin.value_counts()
titanic_df.Cabin.fillna('N', inplace=True)
titanic_df.Cabin.isnull().sum()
titanic_df.Cabin.value_counts()

N              687
G6               4
B96 B98          4
C23 C25 C27      4
D                3
              ... 
D6               1
C7               1
A10              1
B71              1
B102             1
Name: Cabin, Length: 148, dtype: int64

In [22]:
titanic_df.Embarked
titanic_df.Embarked.isnull().sum()
titanic_df.Embarked.fillna('N', inplace=True)
titanic_df.Embarked.isnull().sum()
titanic_df.Embarked.value_counts()

S    644
C    168
Q     77
N      2
Name: Embarked, dtype: int64

In [23]:
from sklearn.preprocessing import LabelEncoder
titanic_df.Cabin = titanic_df.Cabin.str[:1]

le = LabelEncoder()
titanic_df.Cabin = le.fit_transform(titanic_df.Cabin)
print(titanic_df.Cabin)

titanic_df.Sex = le.fit_transform(titanic_df.Sex)
print(titanic_df.Sex)

titanic_df.Embarked = le.fit_transform(titanic_df.Embarked)
print(titanic_df.Embarked)

titanic_df.head()

0      7
1      2
2      7
3      2
4      7
      ..
886    7
887    1
888    7
889    2
890    7
Name: Cabin, Length: 891, dtype: int32
0      1
1      0
2      0
3      0
4      1
      ..
886    1
887    0
888    0
889    1
890    1
Name: Sex, Length: 891, dtype: int32
0      3
1      0
2      3
3      3
4      3
      ..
886    3
887    3
888    3
889    0
890    2
Name: Embarked, Length: 891, dtype: int32


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Cabin  Embarked
0         0       3    1  22.0      1      0   7.2500      7         3
1         1       1    0  38.0      1      0  71.2833      2         0
2         1       3    0  26.0      0      0   7.9250      7         3
3         1       1    0  35.0      1      0  53.1000      2         3
4         0       3    1  35.0      0      0   8.0500      7         3

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Null 처리 함수
def fillna(df):
    df['Age'].fillna(df['Age'].mean(), inplace=True)
    df['Cabin'].fillna('N'n inplace=True)
    df['Embarked'].fillna('N', inplace=True)
    return df

# 머신러닝 알고리즘에 불필요한 속성 제거
def drop_features(df):
    df.drop(['PassengerId', 'Name', 'Ticket'], axis=1, inplace=True)
    return df

# 레이블 인코딩
def format_features(df):
    df['Cabin'] = df['Cabin'].str[:1]
    features = ['Cabin', 'Sex', 'Embarked']
    for feature in features:
        le = LabelEncoder()
        le = le.fit(df[feature])
        df[feature] = le.transform(df[feature])
    return df

# 앞에서 설정한 Data Preprocessing 함수 호출
def transform_features(df):
    df = fillna(df)
    df = drop_features(df)
    df = format_features(df)
    return df